### PR TITLE
Implement feature flag system for flutter tools

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

A configuration setting system based on channel (branch), environment variables, and config settings with defaults and overrides per environment. The objective is to enable a smoother and coordinated rollout of new features within the flutter tool, and to make changes to enable these features easier and more straightforward.

Based on discussion on the intended visibility and release process, I've create four initial feature flags which are related to flutter web, and each of the desktop platforms. This does include a change to the current visibility as it moves all of these to disabled by default. On master channel, each has an associated config that can be enabled, and an environment variable to ensure CI is easy to configure.

This PR does not include any changes to the current behavior yet, but the intention is to remove all of the `FlutterVersion.isMaster` checks and replace them with feature flag checks.

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 1]
```diff
@@ -26,6 +27,26 @@ class ConfigCommand extends FlutterCommand {
       negatable: false,
       hide: !verboseHelp,
       help: 'Print config values as json.');
+    argParser.addFlag(
+      flutterWebFeature.setting.configSetting,
+      help: 'Enable or disable flutter web on master and dev channels.',
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/config.dart]: Drive-by: these help strings seem very likely to get out of sync with the actual channel settings over time. I'd suggest either gating the `addFlag` behind a check that's based on querying availability from the `*Feature` object, or (if you want it available on all channels as in this version) generating an available-channel string programatically from the `Feature`.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/config.dart]: Good point, I've added some logic to generate the help messages from the features themselves. I've also intentionally kept the flags present in all channels to avoid issues with tooling/scripts/users needing to keep track of which flags are defined where

**Comment by jonahwilliams** [comment]
Sorry forgot to push a change before requesting review - just landed change to simplify implementation.

File: packages/flutter_tools/lib/src/base/features.dart [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    }
+    bool isEnabled = featureSetting.enabledByDefault;
+    if (feature.configSetting != null) {
+      final bool configOverride = Config.instance.getValue(feature.configSetting);
+      if (configOverride != null) {
+        isEnabled = configOverride;
+      }
+    }
+    if (feature.environmentOverride != null) {
+      if (platform.environment[feature.environmentOverride] != null) {
```
Comment (blasten) [inline packages/flutter_tools/lib/src/base/features.dart]: nit: would it make sense to check that the value is `'true'`?

File: packages/flutter_tools/lib/src/base/features.dart [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// The [Feature] for Linux desktop.
+const Feature flutterLinuxDesktopFeature = Feature(
+  name: 'Flutter Desktop for Linux',
+  configSetting: 'flutter-linux-desktop',
```
Comment (blasten) [inline packages/flutter_tools/lib/src/base/features.dart]: nit: what about making the `configSetting` the same as the `environmentOverride`?

File: packages/flutter_tools/lib/src/base/features.dart [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+  flutterLinuxDesktopFeature,
+  flutterMacOSDesktopFeature,
+  flutterWindowsDesktopFeature,
+];
+
+/// The [Feature] for flutter web.
+const Feature flutterWebFeature = Feature(
+  name: 'Flutter Web',
+  configSetting: 'flutter-web',
+  environmentOverride: 'FLUTTER_WEB',
```
Comment (blasten) [inline packages/flutter_tools/lib/src/base/features.dart]: nit/minor: `ENABLE_FLUTTER_WEB`

File: packages/flutter_tools/lib/src/base/features.dart [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        break;
+      case 'master':
+      default:
+        featureSetting = feature.master;
+        break;
+    }
+    if (!featureSetting.available) {
+      return false;
+    }
+    bool isEnabled = featureSetting.enabledByDefault;
```
Comment (blasten) [inline packages/flutter_tools/lib/src/base/features.dart]: In the future, it would be nice to be able to control this value remotely. \o/  cc @xster

File: packages/flutter_tools/lib/src/base/features.dart [hunk 6]
```diff
@@ -0,0 +1,245 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/features.dart]: I'm not sure lib/src/base/ is the right place for this file. I think lib/src is probably more appropriate.

File: packages/flutter_tools/lib/src/base/features.dart (hunk 3 above)
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/base/features.dart]: We want to be able to turn on different desktop platforms separately, so they need to be separate settings, but we previously had `ENABLE_FLUTTER_DESKTOP` as an environment variable control so using that means that older third-party instructions about using desktop will continue to work.

File: packages/flutter_tools/lib/src/base/features.dart [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+  bool get isWebEnabled => _isEnabled(flutterWebFeature);
+
+  @override
+  bool get isWindowsEnabled => _isEnabled(flutterWindowsDesktopFeature);
+
+  // Calculate whether a particular feature is enabled for the current channel.
+  static bool _isEnabled(Feature feature) {
+    final String currentChannel = FlutterVersion.instance.channel;
+    FeatureSetting featureSetting;
+    switch (currentChannel) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/features.dart]: Consider making this a map lookup:
```dart
FeatureSetting featureSetting = <String, FeatureSetting>{
  'stable': feature.stable,
  ...
}[currentChannel];
```
and maybe move this lookup to `FeatureSetting`.

File: packages/flutter_tools/lib/src/base/features.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/features.dart]: Similar to https://github.com/flutter/flutter/pull/36138#discussion_r303604941, this environment variable matches the old instructions on how to enable web.

File: packages/flutter_tools/lib/src/base/features.dart (hunk 3 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/base/features.dart]: Got it. What about using `configSetting: 'enable-flutter-desktop'`, so we know that the equivalent env flag is `ENABLE_FLUTTER_DESKTOP`?

Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/base/features.dart]: But with the new approach we want to be able to turn on each desktop platform individually as they stabilize, rather than be stuck with the all-or-nothing of the original environment variable.

Comment (blasten) [reply packages/flutter_tools/lib/src/base/features.dart]: I see.

File: packages/flutter_tools/lib/src/base/features.dart [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    if (feature.environmentOverride != null) {
+      if (platform.environment[feature.environmentOverride] != null) {
+        isEnabled = true;
+      }
+    }
+    return isEnabled;
+  }
+}
+
+/// A [Feature] is process for conditionally enabling tool features.
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/base/features.dart]: is *a* process...

File: packages/flutter_tools/lib/src/base/features.dart [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+///
+/// All settings are optional, and if not provided will generally default to
+/// a "safe" value, such as being off.
+///
+/// The top level feature settings can be provided to apply to all channels.
+/// Otherwise, more specific settings take precidence over higher level
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/base/features.dart]: *precedence*

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/features.dart]: Fixed by removing paragraph (which is no longer relevant).

File: packages/flutter_tools/lib/src/base/features.dart (hunk 8 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/features.dart]: Fixed

File: packages/flutter_tools/lib/src/base/features.dart (hunk 7 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/features.dart]: I moved the lookup to FeatureSetitng, but creating a map involves either losing the safety of the constructor or losing const.

File: packages/flutter_tools/lib/src/base/features.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/features.dart]: Moved

File: packages/flutter_tools/lib/src/base/features.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/features.dart]: Possibly, maybe, speculatively

File: packages/flutter_tools/lib/src/base/features.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/features.dart]: Done

**Comment by jonahwilliams** [comment]
cc @chingjun for google3 specific issues. I don't intend to change any of the feature detection/gating in this PR. We'll likely want to create a google3 specific implementation after the interface rolls.

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 10]
```diff
@@ -26,6 +27,16 @@ class ConfigCommand extends FlutterCommand {
       negatable: false,
       hide: !verboseHelp,
       help: 'Print config values as json.');
+    for (Feature feature in allFeatures) {
```
Comment (xster) [inline packages/flutter_tools/lib/src/commands/config.dart]: You'd want to complete the feedback loop and make these settings actionable by setting session values on google analytics

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/config.dart]: I've wired them up in a comma separated list.

File: packages/flutter_tools/lib/src/usage.dart [hunk 11]
```diff
@@ -54,7 +56,9 @@ const String reloadExceptionTargetPlatform = 'cd27';
 const String reloadExceptionSdkName = 'cd28';
 const String reloadExceptionEmulator = 'cd29';
 const String reloadExceptionFullRestart = 'cd30';
-// Next ID: cd31
+
+const String enabledFlutterFeatures = 'cd31';
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/usage.dart]: If you could take cd32 for this then I'll use cd31 in https://github.com/flutter/flutter/pull/36208

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/usage.dart (hunk 11 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/usage.dart]: Done, next time we'll rowshambo for it :)

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 12]
```diff
@@ -26,6 +27,16 @@ class ConfigCommand extends FlutterCommand {
...
+      if (feature.configSetting == null) {
+        continue;
+      }
+      argParser.addFlag(
+        feature.configSetting,
+        help: feature.generateHelpMessage(),
+        negatable: true,
+      );
+    }
```
Comment (xster) [inline packages/flutter_tools/lib/src/commands/config.dart]: Do we want an extra one to help people get back to default / clear all overrides after they messed with them? Everything turned on isn't going to be the default as we add more experimental features.

File: packages/flutter_tools/lib/src/features.dart [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+
+  @override
+  bool get isWindowsEnabled => _isEnabled(flutterWindowsDesktopFeature);
+
+  // Calculate whether a particular feature is enabled for the current channel.
+  static bool _isEnabled(Feature feature) {
+    final String currentChannel = FlutterVersion.instance.channel;
+    final FeatureSetting featureSetting = feature.getSettingForChannel(currentChannel);
+    if (!featureSetting.available) {
+      return false;
```
Comment (xster) [inline packages/flutter_tools/lib/src/features.dart]: How do we surface this to the user if the user tries to turn something on not available on the channel?

File: packages/flutter_tools/lib/src/features.dart [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+        return dev;
+      case 'master':
+      default:
+        return master;
+    }
+  }
+}
+
+/// A description of the conditions to enable a feature.
+class FeatureSetting {
```
Comment (xster) [inline packages/flutter_tools/lib/src/features.dart]: Are these things ever used independently? Otherwise, just name them FeatureChannelSetting for clarity?

File: packages/flutter_tools/lib/src/features.dart [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+  configSetting: 'enable-windows-desktop',
+  environmentOverride: 'ENABLE_FLUTTER_DESKTOP',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// A [FeatureFlags] that looks up values based on [Feature] definitions.
+class FlutterFeatureFlags implements FeatureFlags {
```
Comment (xster) [inline packages/flutter_tools/lib/src/features.dart]: This class name doesn't add anything to FeatureFlags :)

Perhaps ConfigBasedFeatureFlags?

File: packages/flutter_tools/lib/src/features.dart [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+  /// The settings for the beta branch.
+  final FeatureSetting beta;
+
+  /// The settings for the stable branch.
+  final FeatureSetting stable;
+
+  /// The name of an environment variable that can override the setting.
+  ///
+  /// The environment variable only needs to be "set", that is contain a
+  /// non empty string. This is only intended for usage by CI and not
```
Comment (xster) [inline packages/flutter_tools/lib/src/features.dart]: that's not what you implemented on line 135. Add a test

File: packages/flutter_tools/lib/src/reporting/usage.dart [hunk 17]
```diff
@@ -72,6 +76,17 @@ class Usage {
     _analytics.setSessionValue(kSessionHostOsDetails, os.name);
     // Send the branch name as the "channel".
     _analytics.setSessionValue(kSessionChannelName, flutterVersion.getBranchName(redactUnknownBranches: true));
+    // For each flutter experimental feature, record a session value in a comma
+    // separated list.
+    final String enabledFeatures = allFeatures
+        .where((Feature feature) {
+          return feature.configSetting != null &&
+                 Config.instance.getValue(feature.configSetting) == true;
```
Comment (xster) [inline packages/flutter_tools/lib/src/reporting/usage.dart]: umm we're missing the env vars. Should we just read straight from your _isEnabled implementation? Then we just catch all top level flutter commands. I don't think we set env vars ourselves in reentrant calls anyway.

**Review by xster — COMMENTED** [review]
LGTM

File: packages/flutter_tools/lib/src/reporting/usage.dart (hunk 17 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/reporting/usage.dart]: the env variables are going away though, these are really for our devicelab tests. we want people to use the config.

Comment (xster) [reply packages/flutter_tools/lib/src/reporting/usage.dart]: oki

File: packages/flutter_tools/lib/src/features.dart (hunk 16 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/features.dart]: Done

File: packages/flutter_tools/lib/src/features.dart (hunk 15 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/features.dart]: Good point, removed for now. google3 should be unaffected because we do not use the feature flags yet.

File: packages/flutter_tools/lib/src/features.dart (hunk 14 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/features.dart]: Done

File: packages/flutter_tools/lib/src/features.dart (hunk 13 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/features.dart]: I've added some help text to the config setting if the feature is currently unavailable. The command line help option also includes which channels it can be used with.

File: packages/flutter_tools/lib/src/commands/config.dart (hunk 12 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/config.dart]: Added a `--clear-features` which removes all feature settings.

File: packages/flutter_tools/lib/src/reporting/usage.dart (hunk 17 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/reporting/usage.dart]: Once we've got each of the current set enabled by default on master, I'll remove the corresponding env variable. After these four are gone I'll remove support for the feature entirely.

**Comment by xster** [comment]
Sorry, buyer's remorse. Can you add a usage.dart test in analytics_test.dart too? I think that part is untested.

**Comment by jonahwilliams** [comment]
Will do
